flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -387,35 +365,19 @@ private void stopResources(boolean waitForShutdown) throws InterruptedException
 	 * Add the given stream element queue entry to the operator's stream element queue. This
 	 * operation blocks until the element has been added.
 	 *
-	 * <p>For that it tries to put the element into the queue and if not successful then it waits on
-	 * the checkpointing lock. The checkpointing lock is also used by the {@link Emitter} to output
-	 * elements. The emitter is also responsible for notifying this method if the queue has capacity
-	 * left again, by calling notifyAll on the checkpointing lock.
-	 *
 	 * @param streamElementQueueEntry to add to the operator's queue
 	 * @param <T> Type of the stream element queue entry's result
 	 * @throws InterruptedException if the current thread has been interrupted
 	 */
 	private <T> void addAsyncBufferEntry(StreamElementQueueEntry<T> streamElementQueueEntry) throws InterruptedException {
-		assert(Thread.holdsLock(checkpointingLock));
-
-		pendingStreamElementQueueEntry = streamElementQueueEntry;
-
 		while (!queue.tryPut(streamElementQueueEntry)) {
-			// we wait for the emitter to notify us if the queue has space left again
-			checkpointingLock.wait();
+			yieldToDownstream();
 		}
-
-		pendingStreamElementQueueEntry = null;
 	}
 
-	private void waitInFlightInputsFinished() throws InterruptedException {
-		assert(Thread.holdsLock(checkpointingLock));
-
+	private void waitInFlightInputsFinished() {
 		while (!queue.isEmpty()) {
-			// wait for the emitter thread to output the remaining elements
-			// for that he needs the checkpointing lock and thus we have to free it
-			checkpointingLock.wait();
+			yieldToDownstream();
```

Comment:
Aren't we busy waiting here?

Reply:
Yes, and we have three options to solve that:
* some sleep(1)
* add some mutexes and conditions to be notified when new stuff arrives
* find a cool completablefuture wrapper around the mutexes and conditions (yieldAsync and putAsync and then wait for the first thing to complete)

Which one would you prefer?

Reply:
Couldn't we blockingly wait while yielding? In other words, replacing `tryYield` with `yield`?

Reply:
You are right, now that emitter thread always enqueues results that would work fine.

Reply:
If we do yield now before migrating processing timers, we can run into deadlocks.

Imagine the following: asyncIO is using an external system that is currently not available. The implementation relies on timeouts to occur. Thus, the queue of the AsyncWaitOperator is full with tasks that require the timeout to happen. At that point, we need to release the checkpoint lock instead of yielding (since processor timers are not migrated yet). Thus, we need to alternate between yieldToDownstream and releasing checkpoint lock.

Reply:
AsyncWaitOperator commit will be moved to another PR.